seastar/src/gate.rs

```diff
@@ -88,3 +88,80 @@ impl<'a> GateHolder<'a> {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate as seastar;
+    use futures::join;
+
+    #[seastar::test]
+    async fn test_gate_only_close() {
+        let gate = Gate::new();
+        gate.close().await;
+    }
+
+    #[seastar::test]
+    async fn test_gate_leave_then_close() {
+        let gate = Gate::new();
+
+        {
+            let _ = gate.try_enter().unwrap();
+        }
+
+        gate.close().await;
+    }
+
+    #[seastar::test]
+    async fn test_gate_close_then_leave() {
+        let gate = Gate::new();
+
+        let handler = gate.try_enter().unwrap();
+        let leave_future = async move {
+            drop(handler);
+        };
+
+        let close_future = async {
+            gate.close().await;
+        };
+
+        // join! tries to finish close_future first.
+        join!(close_future, leave_future);
+    }
+
+    #[seastar::test]
+    async fn test_gate_many_leave() {
+        let gate = Gate::new();
+
+        let handler1 = gate.try_enter().unwrap();
+        let leave_future1 = async move {
+            drop(handler1);
+        };
+        let handler2 = gate.try_enter().unwrap();
+        let leave_future2 = async move {
+            drop(handler2);
+        };
+        let handler3 = gate.try_enter().unwrap();
+        let leave_future3 = async move {
+            drop(handler3);
+        };
+
+        let close_future = async {
+            gate.close().await;
+        };
+
+        join!(leave_future1, close_future, leave_future2, leave_future3);
+    }
```

Comment:
Those test cases should verify that the `leave_future{1,2,3}` have completed and dropped their handlers before `gate.close()` resolves.

Reply:
I did that. I also added checking in `test_gate_close_then_leave` that `leave_future` is really polled by `join!` first.